File: PythonLib/CompositePattern.py
import abc
# The classes and/or objects participating in this pattern are:
# 1. Component   (DrawingElement)
# Declares the interface for objects in the composition. Implements
# default behavior for the interface common to all classes, as
# appropriate. declares an interface for accessing and managing its
# child components.
# 2. Leaf   (PrimitiveElement)
# represents leaf objects in the composition. A leaf has no children.
# Defines behavior for primitive objects in the composition.
# 3. Composite   (CompositeElement)
# defines behavior for components having children. Stores child
# components. Implements child-related operations in the Component interface.
# 4. Client  (CompositeApp)
# Manipulates objects in the composition through the Component interface.


# This is the Component, aka 'TreeNode'
class DrawingElement:
    @abc.abstractmethod
    def add(self, d):
        pass

    @abc.abstractmethod
    def remove(self, d):
        pass

    @abc.abstractmethod
    def display(self, indent):
        pass

    @abc.abstractmethod
    def get_name(self):
        pass


# This is 'Leaf'
class PrimitiveElement(DrawingElement):

    def __init__(self, name):
        self.__name = name

    def get_name(self):
        return self.__name

    def add(self, d):
        print("Cannot add to a PrimitiveElement.")

    def remove(self, d):
        print("Cannot remove from a PrimitiveElement.")

    def display(self, indent):
        for i in range(indent):
            print("-", end='')
        print(" " + self.__name)


# This is the 'Composite'
class CompositeElement(DrawingElement):

    def __init__(self, name):
        self.__name = name
        self.__elements = []

    def get_name(self):
        return self.__name

    def add(self, d):
        self.__elements.append(d)

    def remove(self, d):
        self.__elements.remove(d)

    def display(self, indent):
        for i in range(indent):
            print("-", end='')
        print("+ " + self.__name)

        # Display each child element on this node
        for j in range(len(self.__elements)):
            self.__elements[j].display(indent + 2)


# This is the client
def composite_pattern():
    # Create a tree structure
    root = CompositeElement("Picture")
    root.add(PrimitiveElement("Red Line"))
    root.add(PrimitiveElement("Blue Circle"))
    root.add(PrimitiveElement("Green Box"))

    comp = CompositeElement("Two Circles")
    comp.add(PrimitiveElement("Black Circle"))
    comp.add(PrimitiveElement("White Circle"))
    root.add(comp)

    # Add and remove a primitive element
    pe = PrimitiveElement("Yellow Line")
    root.add(pe)
    root.remove(pe)

    # Recursively display nodes
    root.display(1)


if __name__ == "__main__":
    composite_pattern()

File: C++Lib/CompositePattern.cpp
#include <iostream>
#include <vector>
using namespace std;

// The classes and/or objects participating in this pattern are:
// 1. Component   (DrawingElement)
//		Declares the interface for objects in the composition. Implements
//      default behavior for the interface common to all classes, as
//      appropriate. declares an interface for accessing and managing its
//		child components.
// 2. Leaf   (PrimitiveElement)
//		represents leaf objects in the composition. A leaf has no children.
//	    Defines behavior for primitive objects in the composition.
// 3. Composite   (CompositeElement)
//		defines behavior for components having children. Stores child
//		components. Implements child-related operations in the Component interface.
// 4. Client  (CompositeApp)
//		Manipulates objects in the composition through the Component interface.


// This is the "Component". (i.e tree node.)
class DrawingElement {
public:
    virtual void Add(DrawingElement* d) = 0;
    virtual void Remove(DrawingElement* d) = 0;
    virtual void Display(int indent) = 0;
    virtual string getName() = 0;
};

//This is the "Leaf".
class PrimitiveElement : public DrawingElement{
private:
    string name;
public:
    string getName() final {return name;}
    explicit PrimitiveElement(string name) {this->name = move(name);}
    void Add(DrawingElement* c) final {cout << "Cannot add to a PrimitiveElement." << endl;}
    void Remove(DrawingElement* c) final {cout << "Cannot remove from a PrimitiveElement." << endl;}
    void Display(int indent) final {
        for (int i = 1; i <= indent; i++){
            cout << "-";
        }
        cout << " " + name << endl;
    }
};


// This is the "Composite"
class CompositeElement : public DrawingElement{
private:
    string name;
    vector<DrawingElement *> elements;
public:
    string getName() override {return name;}
    explicit CompositeElement(string name) {this->name = move(name);}
    void Add(DrawingElement *d) override {elements.push_back(d);}
    void Remove(DrawingElement *d) override {
        for (int i = 0; i < elements.size(); i++){
            if (elements.at(i)->getName() == d->getName()){
                elements.erase(elements.begin()+i);
                elements.shrink_to_fit();
            }
        }
    }

    void Display(int indent) override {
        for (int i = 1; i <= indent; i++){
            cout << "-";
        }
        cout << "+ " << getName() << endl;

        for (auto & element : elements){
            element->Display(indent + 2);
        }

    }
};

//This is the "client"
int main(){

    //creates a tree structure
    DrawingElement *root;
    root = new CompositeElement("Picture");
    root->Add(new PrimitiveElement("Red Line"));
    root->Add(new PrimitiveElement("Blue Circle"));
    root->Add(new PrimitiveElement("Green Box"));

    DrawingElement *comp;
    comp = new CompositeElement("Two Circles");
    comp->Add(new PrimitiveElement("Black Circle"));
    comp->Add(new PrimitiveElement("White Circle"));
    root->Add(comp);

    // add and remove a PrimitiveElement
    DrawingElement *pe;
    pe = new PrimitiveElement("Yellow Line");
    pe->Add(new PrimitiveElement ("Red Line"));
    root->Add(pe);
    root->Remove(pe);

    // recursively display nodes
    root->Display(1);
}
